learn/shorts/functional-programming.qmd

```diff
@@ -0,0 +1,170 @@
+-   Basic understanding of how functions work
+
+## Necessary software
+
+-   Python editor of choice
+
+## Learning goal
+
+The goal with this learning short is for the reader to understand the basics of functional programming and how it differs from imperative programming. And to be able to answer the following questions:
+
+- What is an immutable data structure?
+- What is a side effect in programming?
+- What is a pure function?
+- What is a higher-order function?
+- How can I use Filter, Map, and Reduce in Python?
+- What does the lambda keyword do in Python?
+- What is list comprehension in Python?
+
+## Take home messages
+
+- Functional programming is a programming paradigm that mainly relies on the concept of immutable data and mathematical functions to perform computation thus trying to avoid side effects.
+
+## Lesson content
+
+### What is an immutable data structure?
+
+An immutable data structure is a data structure that cannot be changed after it is created. In Python, tuples are an example of an immutable data structure. Once a tuple is created, it cannot be changed. This is in contrast to lists, which are mutable. In the context of functional programming, keeping data structures immutable is preferred because they help avoid side effects. The code written should strive to avoid changing the data it is working with. Instead, it should create new data structures based on the existing ones. This makes the code easier to reason about and less prone to bugs.
+
+### What is a side effect in programming?
+
+A side effect is any change in the state of the system that is observable outside the called function other than the return value. This can include modifying a variable outside the function, printing to the console, or modifying a data structure. Side effects can make code harder to understand, as they introduce hidden dependencies and make it difficult to predict the behavior of the code. Functional programming aims to minimize side effects by using immutable data structures and pure functions.
+
+### What is a pure function?
+
+A pure function is a fundamental concept in functional programming, it is basically a function that has no side effects and always returns the same output for the same input. This means that a the ideal pure function does not modify any external state and does not depend on any external state. Truly pure functions are easier to test, reason about, and parallelize, as they do not have hidden dependencies or side effects. They are also easier to reuse, as they can be called with the same arguments and will always return the same result. 
+
+### What is a higher-order function?
+
+A higher-order function is a function which will take one or more existing functions as arguments, and/or a function which returns a new function as its result. Higher-order functions are a powerful concept in functional programming. Examples of higher-order functions in Python include `map`, `filter`, and `reduce`.
```

Comment:
Yea, this first part could be "an intro to functional programming" while the second part could be "Doing functional programming in Python"